src/cargo/ops/cargo_rustc/output_depinfo.rs

```diff
@@ -0,0 +1,128 @@
+use std::borrow::ToOwned;
+use std::collections::HashSet;
+use std::io::{Read, Write};
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+use core::{TargetKind};
+use ops::{Context, Unit};
+use util::{CargoResult, internal, human};
+
+#[derive(Debug)]
+struct DepLine {
+    target: String,
+    deps: Vec<String>,
+}
+
+struct DepFile {
+    dir: String,
+    deps: Vec<DepLine>,
+}
+
+fn render_filename<P: AsRef<Path>>(path: P, basedir: Option<&str>) -> CargoResult<String> {
+    let path = path.as_ref();
+    let relpath = match basedir {
+        None => path,
+        Some(base) => match path.strip_prefix(base) {
+            Ok(relpath) => relpath,
+            _ => path,
+        }
+    };
+    relpath.to_str().ok_or(internal("path not utf-8")).map(ToOwned::to_owned)
+}
+
+fn read_dep_file<P: AsRef<Path>>(path: P) -> CargoResult<DepFile> {
```

Comment:
Could this implementation be shared with the fingerprint module? I think the fingerprint module could store information in a side table in the context perhaps to avoid rereading the filesystem here.